### PR TITLE
Http decompress body when download to file is used

### PR DESCRIPTION
**Pull request by Wavesonics**
This PR addresses some of the issues brought up with this PR: https://github.com/godotengine/godot/pull/48651 via this: https://github.com/godotengine/godot/issues/53085

Downloading to a file with gzip enabled will now automatically unzip the file once the download is complete.

This adds a new function `Compression::decompress_file` that decompresses from one file to another, never loading the whole file into memory.

### CODE REVIEW

**Comment by mhilbrunner** [comment]
Hey! Thanks for contributing.

Note that it may make sense to submit your PR once the work is mostly finished and only work on your own fork and branch until then.
We use drafts more for things which still may change (and shouldn't be merged yet), but are somewhat ready for discussion/review.

This also keeps our backlog a little cleaner, and freshly submitted PRs that are ready to review are usually reviewed faster naturally than drafts that are marked ready for review that have been open for some time.

With that said, you should also squash your commits once the PR is ready. :)

**Comment by Wavesonics** [comment]
right sorry about that @mhilbrunner , things are formatted now and squashed. Should be good to go.

File: scene/main/http_request.cpp [hunk 1]
```diff
@@ -31,6 +31,7 @@
 #include "http_request.h"
 #include "core/io/compression.h"
 #include "scene/main/timer.h"
+#include <core/io/dir_access.h>
```
Comment (mhilbrunner) [inline scene/main/http_request.cpp:34]: Should probably use the same style as the other includes

**Comment by mhilbrunner** [comment]
cc @Faless

**Comment by Faless** [comment]
Hi, first of all, I'm sorry for the very delayed reply. I've been trying to catch up after the holidays.
Please note that the way you are allocating PackedByteArrays is causing a memory leak. See #57122 for how to avoid it.
Regarding the `decompress_*` functions.
Trying to clarify my previous comment, what I would like to see, is something that behave like a stream (in pseudo-code/gdscript):
```gdscript
class Decompressor extends StreamPeer
   # Called to start the stream
   func start(format)
   # Called to decoded partial stream, stores result in local (instance) ring buffer
   func put_partial_data(PackedByteArray/ptr+size) -> Error # (OK, FAIL)
   # Size of decompressed bytes in local buffer
   func get_available_bytes() -> int
   # Called to retrieve partially decoded bytes
   func get_partial_data(int/ptr+size) -> [PackedByteArrray, Error]/Error # (OK, FAIL)

   # Config
   var max_buffer_size : int
   var chunk_size : int
```

**Comment by Wavesonics** [comment]
@Faless Ah I see what you mean now.

So I fixed the memory leak, and also fixed two edge cases around the max body size being exceeded, so it's now reported correctly, and never prematurely truncated.

I like the idea of the `Decompressor` class, I'm thinking it might make more sense as a separate PR. And maybe another PR after that which would convert these functions over to using it.

**Comment by Wavesonics** [comment]
Updated this to fix merge conflicts, and I started another branch (based off this one) to begin work on the Gzip stream class

**Comment by akien-mga** [comment]
Superseded by #63740.

**Pull request closed without merging**